tests/core/test_configuration.py

```diff
@@ -940,3 +785,384 @@ def test_gettimedelta(self):
+
+    @pytest.mark.parametrize(
+        "conf_dict",
+        [
+            {},  # Even if the section is absent from config file, environ still needs replacing.
+            {'core': {'hostname_callable': 'socket:getfqdn'}},
+        ],
+    )
+    def test_deprecated_values_from_environ(self, conf_dict):
+        def make_config():
+            test_conf = AirflowConfigParser(
+                default_config="""
+[core]
+executor=SequentialExecutor
+[database]
+sql_alchemy_conn=sqlite://test
+"""
+            )
+            # Guarantee we have a deprecated setting, so we test the deprecation
+            # lookup even if we remove this explicit fallback
+            test_conf.deprecated_values = {
+                'core': {'hostname_callable': (re.compile(r':'), r'.', '2.1')},
+            }
+            test_conf.read_dict(conf_dict)
+            test_conf.validate()
+            return test_conf
+
+        with pytest.warns(FutureWarning):
+            with unittest.mock.patch.dict('os.environ', AIRFLOW__CORE__HOSTNAME_CALLABLE='socket:getfqdn'):
+                test_conf = make_config()
+                assert test_conf.get('core', 'hostname_callable') == 'socket.getfqdn'
+
+        with reset_warning_registry():
+            with warnings.catch_warnings(record=True) as warning:
+                with unittest.mock.patch.dict(
+                    'os.environ',
+                    AIRFLOW__CORE__HOSTNAME_CALLABLE='CarrierPigeon',
+                ):
+                    test_conf = make_config()
+                    assert test_conf.get('core', 'hostname_callable') == 'CarrierPigeon'
+                    assert [] == warning
+
+    def test_deprecated_funcs(self):
+        for func in [
+            'load_test_config',
+            'get',
+            'getboolean',
+            'getfloat',
+            'getint',
+            'has_option',
+            'remove_option',
+            'as_dict',
+            'set',
+        ]:
+            with mock.patch(f'airflow.configuration.conf.{func}') as mock_method:
+                with pytest.warns(DeprecationWarning):
+                    getattr(configuration, func)()
+                mock_method.assert_called_once()
+
+    @pytest.mark.parametrize("display_source", [True, False])
+    @mock.patch.dict('os.environ', {}, clear=True)
+    def test_conf_as_dict_when_deprecated_value_in_config(self, display_source: bool):
+        with use_config(config="deprecated.cfg"):
+            # test display_source
+            cfg_dict = conf.as_dict(
+                display_source=display_source, display_sensitive=True, include_env=False, include_cmds=False
+            )
+            assert cfg_dict['core'].get('sql_alchemy_conn') == (
+                ('mysql://', "airflow.cfg") if display_source else 'mysql://'
+            )
+            # database should be None because the deprecated value is set in config
+            assert cfg_dict['database'].get('sql_alchemy_conn') is None
```

Comment:
i am wondering whether we need to add a follow up test like:

```
conf.read_dict(conf_dict, source=cfg_path)
assert conf.read('database', 'sql_alchemy_conn`) == conf.read('core', 'sql_alchemy_conn`)
```

```

https://github.com/apache/airflow/blob/de3c0389eb42fa59eb9a4ad9caa2cd85baa367a0/airflow/cli/commands/task_command.py#L345

Reply:
Not really. If someone sets both values and they are different, the "database" one should take precedence. 

We basically do not care what the "core" value is once "database" is set. but if "database" is not set (which will be default for anyone coming from 2.2.*) then the "core" value should be returned when "database" is used rather than default.

Reply:
The change I implemented, does not change the "get" behaviour. This was working fine before. Nowhere in airflow we use "core" any more - we always retrieve the "database" entry, so we basically do not care whate get("core") returns - and we do not want to do anything with it. The conf.get("core", "sql_alchemy_conn") is not supposed to return a "good" value any more - we don't care about it any more as this is an internal detail only.

The problem was only when "as_dict" was called when "database" was not set - then it used "default" value for "database" rather than corresponding "core" values - which was the root cause of the problem.

Reply:
ah, i made a mistake, sorry for the confusion. I meant to say:

```
conf.read_dict(conf_dict, source=cfg_path)
assert conf.read('database', 'sql_alchemy_conn`) == "the correct value of sql_alchemy_conn"
```

you are right, we don't need to care about that value in the `core`.

so adding a check after the `read_dict` from the dict dumped by `as_dict` can ensure the correct behavior since our code does not access the `dict` directly.

Reply:
I think this is more-or-less what the unit tests do for all the "as-dict" variants (but they go a a bit deeper i.e. they actually check whether both values are as expected, but yeah - why not - we can also make conf.read() test - since I added context managers that save/restore the conf, this would even have no side-effects. Let me add it :)

Reply:
Indeed - that made the tests a bit more explicit - showing the final "goal" we want to achieve while we are testing it :). 

I also made sure to delete all env vars before we run conf.get() and it nicely shows that "run_as_user" will also work fine with the conf where env vars are not passed  - so that SECRETS/CMD/ENV vars are passed in the dict rather than via envs.

Good idea to add those asserts @pingzh :).